### ISSUE TITLE
Rename uploaded files with UUID

### ISSUE THREAD

**Issue by tylerecouture** [bytedeck/bytedeck#767]
Thanks to @aidenmitchell:

Currently, the files aren't very well hidden and could be easy to guess, for example:

https://d10ge8y4vx8iud.cloudfront.net/public_media/documents/2020/12/09/Capture.PNG

Changing the date could easily find other screen captures with the same filename.  Instead we should rename files with a UUID, similar to how summernote widget does when you embed an image:

https://d10ge8y4vx8iud.cloudfront.net/public_media/django-summernote/2016-10-15/09212b01-57da-4058-b6cd-c2e68c189eb6.png

**Comment by thefoxwithin** [comment]
I've explored a couple of approaches, both of which involve implementing custom `upload_to` functions since Django's default behavior isn't flexible enough to cleanly support UUIDs out of the box.

### **Option 1: A single generic `upload_to_uuid` function**

This version generates a UUID-based filename and saves everything in a single upload directory (e.g. `_media_uploads`).

**Pros:**
- Simple and easy to implement across all models
- Adds the UUID to filenames as intended

**Cons:**
- All uploaded files go into one directory
- No subfolder structure, which means no sorting or visual distinction by context (e.g. badges vs icons)

---

### **Option 2: Multiple specialized `upload_to_*` functions**

This approach defines a separate function per upload context (e.g. `upload_to_badge_icon`, `upload_to_quest_image`, etc.), each pointing to its own subfolder like `_media_uploads/icons/badges/`.

**Pros:**
- Adds the UUID to filenames
- Maintains existing folder structure, so files remain logically grouped and sortable

**Cons:**
- Adds a lot of boilerplate — potentially many separate functions in `utilities/models.py`
- Requires creating a new `upload_to_*` function each time a new upload path is needed, which increases maintenance overhead

---

I'm currently trying to find a third option that combines the **simplicity** of the first approach with the **structured organization** of the second — ideally something that keeps the original `upload_to` behavior but simply adds a UUID to the filename.

One possible solution I’ve found is [`django-file-upload-to`](https://github.com/panttojo/django-file-upload-to/), which looks promising. However, I ran into some issues with `makemigrations` when trying to use it, so I'm still digging into that.

Will keep exploring and update again when I have something more stable.

**Comment by tylerecouture** [comment]
Regarding option 1, "All uploaded files go into one directory" <-- this is not true, because we can set the upload_to parameter to a function, and we can do whatever we want in the function, right?

See the docs: https://docs.djangoproject.com/en/4.2/ref/models/fields/#django.db.models.FileField.upload_to

> **upload_to** may also be a callable, such as a function. This will be called to obtain the upload path, including the filename...


Some additional notes on this:

Code for files uploaded with comments (which is the main concern here) currently uses the model at `comments.models.Document`  that has a FileField and a foreign key to a Comment object.

The files themselves are uploaded as part of the POST request, and the different views are where the files are saved into the our Document model, for example in the function-based-view quest_manager.views.complete() we have:

```
    # Create Document objects and connect them to the comment for uploaded files
    if request.FILES:
        for afile in request.FILES.getlist("attachments"):
            newdoc = Document(docfile=afile, comment=draft_comment)
            newdoc.full_clean()
            newdoc.save()
```

And then for teachers we have the class-based-view quest_manager.views.ApproveView and the save_uploaded_files() method that uses identical code.

While we're working on files, we should combine that code into a common function.

**Comment by thefoxwithin** [comment]
> Regarding option 1, "All uploaded files go into one directory" <-- this is not true, because we can set the upload_to parameter to a function, and we can do whatever we want in the function, right?
> 
> See the docs: https://docs.djangoproject.com/en/4.2/ref/models/fields/#django.db.models.FileField.upload_to
> 
> > **upload_to** may also be a callable, such as a function. This will be called to obtain the upload path, including the filename...
> 

You're correct. For option 1, I meant creating a single `upload_to_uuid` function, which would not have any dynamic behavior. This means that every file uploaded using `ImageField(upload_to=upload_to_uuid, ...)` would end up in the same directory.

The example from the documentation is similar to option 2. However, instead of defining `upload_to_uuid_directory_path` in `.utilities/models.py`, each `models.py` that uses the function would define its own `upload_to_uuid_directory_path`. So, there would still be multiple uuid-related functions, but they’d be defined locally within each relevant `models.py` file, rather than in a shared utilities file.

I first tried making a singular `upload_to_uuid` that had the option to add in a string so that it would mimic the original `upload_to` but just add a uuid. Which is how I came across not being able to `makemigrations` I'm still looking for a solution that uses this method, but if we only need it in a specific area, I can make the single `upload_to_uuid` with the right path in the `comments/models.py` file.

Let me know if anything is unclear or if you’d like more details!

> Some additional notes on this:
> 
> Code for files uploaded with comments (which is the main concern here) currently uses the model at `comments.models.Document` that has a FileField and a foreign key to a Comment object.
> 
> The files themselves are uploaded as part of the POST request, and the different views are where the files are saved into the our Document model, for example in the function-based-view quest_manager.views.complete() we have:
> 
> ```
>     # Create Document objects and connect them to the comment for uploaded files
>     if request.FILES:
>         for afile in request.FILES.getlist("attachments"):
>             newdoc = Document(docfile=afile, comment=draft_comment)
>             newdoc.full_clean()
>             newdoc.save()
> ```
> 
> And then for teachers we have the class-based-view quest_manager.views.ApproveView and the save_uploaded_files() method that uses identical code.
> 
>While we're working on files, we should combine that code into a common function.

I've made that change in my code so now both use a helper function with documentation for it. Since both the views already have pretty extensive tests that cover the usage of the helper function adding tests for the helper function itself carries little meaning.
